### PR TITLE
[#94] updated code to support devise 4+

### PR DESCRIPTION
**Pull request by swapnilchincholkar**
- updated test cases
- tested code using building gem for devise 4.0.1 & 4.2.0

### CODE REVIEW

**Comment by baschtl** [comment]
@swapnilchincholkar Thanks for your contribution! I will have a look at this within the next days.

**Comment by Geesu** [comment]
@baschtl Just checking to see if you've had a chance to look at this?

File: test/devise/async/model_test.rb [hunk 1]
```diff
@@ -3,59 +3,79 @@
 module Devise
   module Async
     describe "Model" do
+
+      # As per the https://github.com/plataformatec/devise/blob/master/CHANGELOG.md#410,
```
Comment (baschtl) [inline test/devise/async/model_test.rb]:
```
# As per https://github.com/plataformatec/devise/blob/master/CHANGELOG.md#410,
# within a transaction mail will be enqueued and after commit send_devise_pending_notifications will be invoked
```

File: test/devise/async/model_test.rb [hunk 2]
```diff
@@ -3,59 +3,79 @@
...
+      # within transaction mail will be enqued & after commit send_devise_pending_notifications will be invoked
+
+      # 1. Within transaction i.e object creation, mails will only be accummulated
```
Comment (baschtl) [inline test/devise/async/model_test.rb]:
```
# 1. Within a transaction, i.e. object creation, mails will only be accummulated

File: test/devise/async/model_test.rb [hunk 3]
```diff
@@ -3,59 +3,79 @@
...
+
+      # As per the https://github.com/plataformatec/devise/blob/master/CHANGELOG.md#410,
+      # within transaction mail will be enqued & after commit send_devise_pending_notifications will be invoked
+
+      # 1. Within transaction i.e object creation, mails will only be accummulated 
       it "accumulates notifications to be sent after commit on Model creation" do
+        admin = create_admin
         Admin.transaction do
-          admin = create_admin
           mailers = admin.send(:devise_pending_notifications) # [:confirmation_instructions, ["RUQUib67wLcCiEyZMwfx", {}]]
```
Comment (baschtl) [inline test/devise/async/model_test.rb]: The comment in this line doesn't make sense if `mailers` is empty, right?

File: test/devise/async/model_test.rb [hunk 4]
```diff
@@ -3,59 +3,79 @@
...
+        end  
+        admin[:username] = "changed_username"
+        admin.send_confirmation_instructions
 
-          mailer = mailers.first
-          mailer.size.must_equal 2
-          mailer.first.must_equal :confirmation_instructions
-          mailer.last.must_be_instance_of Array
-        end
+        mailers = admin.send(:devise_pending_notifications) # [[:confirmation_instructions, ["RUQUib67wLcCiEyZMwfx", {}]]]
```
Comment (baschtl) [inline test/devise/async/model_test.rb]: Could you rename `mailers` to `notifications` and `mailer` to `notification`? I find the naming to be quite confusing.

File: test/devise/async/model_test.rb [hunk 5]
```diff
@@ -3,59 +3,79 @@
...
+        Worker.expects(:enqueue).never # after_commit will not fire without save
       end
 
+      # 2. On explicit save, after commit will be invoked hence notifications will be queued
       it "immediately sends notifications when the model has not changed" do
-        admin = create_admin
-        Worker.expects(:enqueue).with(:confirmation_instructions, "Admin", admin.id.to_s, instance_of(String), {})
-        admin.send_confirmation_instructions
+        Admin.transaction do
+          @admin2 = create_admin
```
Comment (baschtl) [inline test/devise/async/model_test.rb]: Why does this need to be an instance variable? I guess `admin` would be a more suitable name here as there is only one admin been created.

File: test/devise/async/model_test.rb [hunk 6]
```diff
@@ -3,59 +3,79 @@
...
 
-          Worker.expects(:enqueue).never # after_commit will not fire without save
+          mailers.size.must_equal 2         
+          notification_type.must_equal :confirmation_instructions
+          mailers.last.must_be_instance_of Array
+          Worker.expects(:enqueue).with(:confirmation_instructions, "Admin", admin.id.to_s, instance_of(String), {})
         end
       end
 
+      # 4. As data changed & saved, i.e after_commit is invoked hence notifications are queued to worker
```
Comment (baschtl) [inline test/devise/async/model_test.rb]:
```
# 4. As data changed and has been saved, i.e. after_commit is invoked, notifications are queued to worker
```

File: test/devise/async/model_test.rb [hunk 7]
```diff
@@ -3,59 +3,79 @@
...
 
-          mailer = mailers.first
-          mailer.size.must_equal 2
-          mailer.first.must_equal :confirmation_instructions
-          mailer.last.must_be_instance_of Array
-
-          admin.save
-          Worker.expects(:enqueue).with(:confirmation_instructions, "Admin", admin.id.to_s, instance_of(String), {})
+          admin.save                     # after_commit will be invoked
+          2.times do
```
Comment (baschtl) [inline test/devise/async/model_test.rb]: This construct looks weird. Why haven't you used something like this:
```ruby
Worker.expects(:enqueue).with(:confirmation_instructions, "Admin", admin.id.to_s, instance_of(String), {})
admin.save
```
?

I am not that used to minitest. But I thought that you first set the expectation and then run the code that triggers it. If the code triggers it before the end of the example the expectation is being satisfied. Is this how minitest works here?

File: test/devise/async/model_test.rb [hunk 8]
```diff
@@ -81,18 +101,24 @@ module Async
         after { I18n.locale = nil }
 
         it "should set the current locale to the args" do
-          admin = create_admin
-          Worker.expects(:enqueue).with(:confirmation_instructions, "Admin", admin.id.to_s, instance_of(String), { 'locale' => :de })
-          admin.send_confirmation_instructions
+          Admin.transaction do
```
Comment (baschtl) [inline test/devise/async/model_test.rb]: I don't get why this example and the next one you adapted look so different. The only difference should be the (non-) existence of the locale in the args.

**Comment by baschtl** [comment]
Closing this as there has not been any reaction to the code review for months and the latest [release](https://github.com/mhfs/devise-async/releases/tag/v1.0.0) depends on ActiveJob. The code changed dramatically by now.

**Pull request closed without merging**